presto-main/src/main/java/io/prestosql/server/HttpRequestSessionContext.java

```diff
@@ -125,9 +127,21 @@ public HttpRequestSessionContext(HeaderSupport forwardedHeaderSupport, HttpServl
                     .build());
 
         String user = trimEmptyToNull(servletRequest.getHeader(PRESTO_USER));
+
         if (user == null) {
             user = authenticatedUser;
         }
+        originalIdentity = Optional.ofNullable(user)
+                .map(prestoUser -> Identity.forUser(prestoUser)
+                    .withPrincipal(Optional.ofNullable(servletRequest.getUserPrincipal()))
+                    .withRoles(parseRoleHeaders(servletRequest))
+                    .withExtraCredentials(parseExtraCredentials(servletRequest))
+                    .build());
+        String sessionAuthorizationUsername = trimEmptyToNull(servletRequest.getHeader(PRESTO_SET_SESSION_AUTHORIZATION_USERNAME));
+
+        if (sessionAuthorizationUsername != null) {
+            user = sessionAuthorizationUsername;
```

Comment:
We should preserve `user` here and store `sessionAuthorizationUsername` separately, as we'll need to check that both the authorization user can impersonate `user` can impersonate `sessionAuthorizationUsername`.

Example:

authenticated user: `a`
X-Presto-User: `b`
X-Set-Session-Authorization-User: `c`

We need to call `canImpersonate(a, b)` and `canImpersonate(b, c)`. We also need to call the legacy `checkCanSetUser(principal, b)` and `checkCanSetUser(principal, c)`.

@dain has several outstanding refactoring covering this, so please wait a week for those to land.

Reply:
It looks like `QuerySessionSupplier` already does `canImpersonate(a, b)` and `checkCanSetUser(principal, b)`. Seems like I'll have to add `canImpersonate(b, c)` and `checkCanSetUser(principal, c)` to `SetSessionAuthorizationTask`.

I set the user to sessionAuthorizationUsername so that the set session authorization username can be propagated to the session later. If we keep `user` separate from `sessionAuthorizationUsername`, how do you envision `sessionAuthorizationUsername` getting into the session identity field, which uses `context.getIdentity`?

something like this in the QuerySessionSupplier class?
```
Identity setSessionAuthorizationIdentity = context.getSetSessionAuthorizationIdentity();
                                                           
SessionBuilder sessionBuilder = Session.builder(sessionPropertyManager)
                .setQueryId(queryId)
                .setIdentity(setSessionAuthotizationIdentity != null ? setSessionAuthotizationIdentity : identity)
```

Reply:
I want to ask if `checkCanSetUser(principal, c)` is needed. It seems like this method has been deprecated and replaced with `checkCanImpersonateUser`. `checkCanSetUser(principal, b)` is left for backwards compatibility but assuming a chain of impersonation is what we want (a can impersonate b, and then b can impersonate c), would only 3 checks be sufficient?